File: src/detect/node.ts
import { join } from 'node:path';
import { exists, readJson } from '../utils/fs.js';

interface PackageJson {
	dependencies?: Record<string, string>;
	devDependencies?: Record<string, string>;
}

type PartialStack = Omit<import('./index.js').DetectedStack, 'features'>;

function hasDep(pkg: PackageJson, name: string): boolean {
	return Boolean(pkg.dependencies?.[name] || pkg.devDependencies?.[name]);
}

export async function detectNode(cwd: string): Promise<PartialStack | null> {
	const pkgPath = join(cwd, 'package.json');
	if (!(await exists(pkgPath))) return null;

	const pkg = await readJson<PackageJson>(pkgPath);

	const result: PartialStack = {
		language: hasDep(pkg, 'typescript') || (await exists(join(cwd, 'tsconfig.json')))
			? 'typescript'
			: 'javascript',
		framework: null,
		preset: null,
		testRunner: null,
		linter: null,
		typeChecker: null,
		formatter: null,
		packageManager: null,
	};

	// Framework detection
	if (hasDep(pkg, '@sveltejs/kit')) {
		result.framework = 'sveltekit';
		result.preset = 'sveltekit-ts';
	} else if (hasDep(pkg, 'next')) {
		result.framework = 'next';
		result.preset = 'react-next-ts';
	} else if (hasDep(pkg, 'nuxt')) {
		result.framework = 'nuxt';
		result.preset = 'vue-nuxt-ts';
	} else if (hasDep(pkg, 'vue')) {
		result.framework = 'vue';
		result.preset = 'vue-nuxt-ts';
	} else if (hasDep(pkg, 'express')) {
		result.framework = 'express';
		result.preset = 'node-express';
	} else if (hasDep(pkg, 'fastify')) {
		result.framework = 'fastify';
		result.preset = 'node-express';
	}

	// Test runner
	if (hasDep(pkg, 'vitest')) {
		result.testRunner = { name: 'vitest', command: 'npx vitest run' };
	} else if (hasDep(pkg, 'jest')) {
		result.testRunner = { name: 'jest', command: 'npx jest' };
	} else if (hasDep(pkg, 'mocha')) {
		result.testRunner = { name: 'mocha', command: 'npx mocha' };
	}

	// Linter
	if (hasDep(pkg, 'eslint')) {
		result.linter = { name: 'eslint', command: 'npx eslint .' };
	} else if (hasDep(pkg, '@biomejs/biome')) {
		result.linter = { name: 'biome', command: 'npx biome check .' };
	}

	// Type checker
	if (await exists(join(cwd, 'tsconfig.json'))) {
		if (result.framework === 'sveltekit') {
			result.typeChecker = { name: 'svelte-check', command: 'npm run check' };
		} else if (result.framework === 'next') {
			result.typeChecker = { name: 'tsc', command: 'npx tsc --noEmit' };
		} else {
			result.typeChecker = { name: 'tsc', command: 'npx tsc --noEmit' };
		}
	}

	// Formatter
	if (hasDep(pkg, 'prettier')) {
		result.formatter = { name: 'prettier', command: 'npx prettier --write' };
	} else if (hasDep(pkg, '@biomejs/biome')) {
		result.formatter = { name: 'biome', command: 'npx biome format --write .' };
	}

	// Package manager
	if (await exists(join(cwd, 'pnpm-lock.yaml'))) {
		result.packageManager = 'pnpm';
	} else if (await exists(join(cwd, 'yarn.lock'))) {
		result.packageManager = 'yarn';
	} else if (await exists(join(cwd, 'bun.lockb'))) {
		result.packageManager = 'bun';
	} else {
		result.packageManager = 'npm';
	}

	return result;
}

File: internal/detect/node.go
package detect

import (
	"encoding/json"
	"github.com/samahlstrom/forge-cli/internal/util"
	"path/filepath"
)

type packageJSON struct {
	Dependencies    map[string]string `json:"dependencies"`
	DevDependencies map[string]string `json:"devDependencies"`
}

func hasDep(pkg packageJSON, name string) bool {
	if _, ok := pkg.Dependencies[name]; ok {
		return true
	}
	if _, ok := pkg.DevDependencies[name]; ok {
		return true
	}
	return false
}

// DetectNode detects Node.js/TypeScript projects.
func DetectNode(cwd string) *DetectedStack {
	pkgPath := filepath.Join(cwd, "package.json")
	if !util.Exists(pkgPath) {
		return nil
	}

	data, err := util.ReadText(pkgPath)
	if err != nil {
		return nil
	}
	var pkg packageJSON
	if err := json.Unmarshal([]byte(data), &pkg); err != nil {
		return nil
	}

	lang := "javascript"
	if hasDep(pkg, "typescript") || util.Exists(filepath.Join(cwd, "tsconfig.json")) {
		lang = "typescript"
	}

	result := &DetectedStack{Language: lang}

	// Framework
	switch {
	case hasDep(pkg, "@sveltejs/kit"):
		result.Framework = "sveltekit"
		result.Preset = "sveltekit-ts"
	case hasDep(pkg, "next"):
		result.Framework = "next"
		result.Preset = "react-next-ts"
	case hasDep(pkg, "nuxt"):
		result.Framework = "nuxt"
		result.Preset = "vue-nuxt-ts"
	case hasDep(pkg, "vue"):
		result.Framework = "vue"
		result.Preset = "vue-nuxt-ts"
	case hasDep(pkg, "express"):
		result.Framework = "express"
		result.Preset = "node-express"
	case hasDep(pkg, "fastify"):
		result.Framework = "fastify"
		result.Preset = "node-express"
	case hasDep(pkg, "hono"):
		result.Framework = "hono"
		result.Preset = "react-next-ts"
	case hasDep(pkg, "drizzle-orm"):
		result.Framework = "node-api"
		result.Preset = "react-next-ts"
	case hasDep(pkg, "@trpc/server"):
		result.Framework = "node-api"
		result.Preset = "react-next-ts"
	}

	// Fallback: if we detected language + tools but no framework, still assign a preset
	if result.Preset == "" {
		if lang == "typescript" {
			result.Preset = "react-next-ts"
		} else {
			result.Preset = "react-next-ts"
		}
	}

	// Test runner
	switch {
	case hasDep(pkg, "vitest"):
		result.TestRunner = &ToolInfo{"vitest", "npx vitest run"}
	case hasDep(pkg, "jest"):
		result.TestRunner = &ToolInfo{"jest", "npx jest"}
	case hasDep(pkg, "mocha"):
		result.TestRunner = &ToolInfo{"mocha", "npx mocha"}
	}

	// Linter
	switch {
	case hasDep(pkg, "eslint"):
		result.Linter = &ToolInfo{"eslint", "npx eslint ."}
	case hasDep(pkg, "@biomejs/biome"):
		result.Linter = &ToolInfo{"biome", "npx biome check ."}
	}

	// Type checker
	if util.Exists(filepath.Join(cwd, "tsconfig.json")) {
		if result.Framework == "sveltekit" {
			result.TypeChecker = &ToolInfo{"svelte-check", "npm run check"}
		} else {
			result.TypeChecker = &ToolInfo{"tsc", "npx tsc --noEmit"}
		}
	}

	// Formatter
	switch {
	case hasDep(pkg, "prettier"):
		result.Formatter = &ToolInfo{"prettier", "npx prettier --write"}
	case hasDep(pkg, "@biomejs/biome"):
		result.Formatter = &ToolInfo{"biome", "npx biome format --write ."}
	}

	// Package manager
	switch {
	case util.Exists(filepath.Join(cwd, "pnpm-lock.yaml")):
		result.PackageManager = "pnpm"
	case util.Exists(filepath.Join(cwd, "yarn.lock")):
		result.PackageManager = "yarn"
	case util.Exists(filepath.Join(cwd, "bun.lockb")):
		result.PackageManager = "bun"
	default:
		result.PackageManager = "npm"
	}

	return result
}
